### ISSUE TITLE
word_tokenize not tokenizing strings containing ','

### ISSUE THREAD

**Issue by satyammittal** [nltk/nltk#1963]
I tried
```
from nltk.tokenize import word_tokenize
a="g, a, b, c, 123, g32,12 123121 {1}"
word_tokenize(a)
```
**Output I am getting:** 
['g', ',', 'a', ',', 'b', ',', 'c', ',', '123', ',', 'g32,12', '123121', '{', '1', '}']
**Output should be**
['g', ',', 'a', ',', 'b', ',', 'c', ',', '123', ',', 'g32',',','12', '123121', '{', '1', '}']

**Comment by beniza** [comment]
word_tokenize works only at the sentence level. So you'll have to split at the sentence level and the tokenize the sentences.
You can remove punctuation without using NLTK. 
`Please ref. to this question on SE for more info:` https://stackoverflow.com/questions/15547409/how-to-get-rid-of-punctuation-using-nltk-tokenizer

**Comment by jnothman** [comment]
Are you concerned about "g32,12" staying as one token? Yes, this is in accordance with the rules that word_tokenize follows, and is consistent with other implementations of Penn TreeBank-style tokenization. Consistency across your NLP pipeline is often more important than the specific rules.

**Comment by alvations** [comment]
I'm interested in which regex causes that particular "non-split" though. I suspect it's the digits punctuation regex keeping the comma between numbers. It'll be good to know. Get back to you all in a while.

**Comment by alvations** [comment]
Yes, it looks like the [digits punctuation regex](https://github.com/nltk/nltk/blob/develop/nltk/tokenize/treebank.py#L75) is forcing the comma between the numbers. 

```python
from nltk.tokenize import TreebankWordTokenizer

tbw = TreebankWordTokenizer()
text = "g, a, b, c, 123, g32,12 123121 {1}"

# No change after these substitution.
for regexp, substitution in tbw.STARTING_QUOTES:
    text = regexp.sub(substitution, text)

# Change comes in the 2nd regex. 
for regexp, substitution in tbw.PUNCTUATION:
    text = regexp.sub(substitution, text)
    print(regexp)
    print(text)
    print()
```

[out]:

```

re.compile('([^\\.])(\\.)([\\]\\)}>"\\\'»”’ ]*)\\s*$')
g, a, b, c, 123, g32,12 123121 {1}

re.compile('([:,])([^\\d])')
g ,  a ,  b ,  c ,  123 ,  g32,12 123121 {1}

re.compile('([:,])$')
g ,  a ,  b ,  c ,  123 ,  g32,12 123121 {1}

re.compile('\\.\\.\\.')
g ,  a ,  b ,  c ,  123 ,  g32,12 123121 {1}

re.compile('[;@#$%&]')
g ,  a ,  b ,  c ,  123 ,  g32,12 123121 {1}

re.compile('([^\\.])(\\.)([\\]\\)}>"\\\']*)\\s*$')
g ,  a ,  b ,  c ,  123 ,  g32,12 123121 {1}

re.compile('[?!]')
g ,  a ,  b ,  c ,  123 ,  g32,12 123121 {1}

re.compile("([^'])' ")
g ,  a ,  b ,  c ,  123 ,  g32,12 123121 {1}
```


But keeping the character before the digits is not well-motivated since we usually get text that contains `US$10,000` and expect the tokenizer to return `['US', '$' , '10,000']`.  But if we have something like `US10,000` this is sort of an unexpected behavior: 

```python
from nltk.tokenize import TreebankWordTokenizer
tbw = TreebankWordTokenizer()
text = "I have US10,000 in the bank, you have none."

for regexp, substitution in tbw.STARTING_QUOTES:
    text = regexp.sub(substitution, text)
    
for regexp, substitution in tbw.PUNCTUATION:
    text = regexp.sub(substitution, text)
    print(regexp)
    print(text)
    print()
```

[out]:
```
re.compile('([^\\.])(\\.)([\\]\\)}>"\\\'»”’ ]*)\\s*$')
I have US10,000 in the bank, you have none .  

re.compile('([:,])([^\\d])')
I have US10,000 in the bank ,  you have none .  

re.compile('([:,])$')
I have US10,000 in the bank ,  you have none .  

re.compile('\\.\\.\\.')
I have US10,000 in the bank ,  you have none .  

re.compile('[;@#$%&]')
I have US10,000 in the bank ,  you have none .  

re.compile('([^\\.])(\\.)([\\]\\)}>"\\\']*)\\s*$')
I have US10,000 in the bank ,  you have none  . 

re.compile('[?!]')
I have US10,000 in the bank ,  you have none  . 

re.compile("([^'])' ")
I have US10,000 in the bank ,  you have none  . 
```

**Comment by jnothman** [comment]
Oh of course, it's only when the comma is followed by a digit that it is
not split: '([:,])([^\d])' -> ' \1 \2'​

Shrug. It's too rare a case to handle specially, isn't it?

The rule could instead be '(\b[\d+][:,])([^\d])' but you really need a whole
lot of non-regression tests to make any change to a tokeniser.

**Comment by jnothman** [comment]
Sorry, wrong expression. Hard to compose. Hard to test. Hard to maintain. I'd say Won't Fix.

**Comment by alvations** [comment]
Yes, if we do try to fix this, it's hard to maintain and not the true port of the sed `TreebankWordTokenizer`. 

The [chervon fix](https://github.com/nltk/nltk/blob/develop/nltk/tokenize/__init__.py#L97) is an easy hack on top of it. And maybe these hacks has to been handled in a similar way. 

But I do know of one type of counter examples where we would want the commas to be kept when the "token" is alphanumeric, e.g. [`propane-1,1,1-triol`](https://www.reddit.com/r/askscience/comments/1s93gp/what_are_the_commaseparated_lists_of_numbers_in/). With the existing `word_tokenize`, it does tokenize such tokens perfectly and personally I do like to keep this "featture/bug" =)

**Comment by alvations** [comment]
@satyammittal Thank you for raising this bug/feature! 

I suggest chaining tokenizers to achieve your desired output. Hope this helps:

```python
from functools import partial

from nltk import word_tokenize

punct_tokenize = partial(re.sub, pattern='([.,!?(){}]+)', repl=' \g<1> ', flags=re.U)

text = "g, a, b, c, 123, g32,12 123121 {1}"
expected = ['g', ',', 'a', ',', 'b', ',', 'c', ',', '123', ',', 'g32',',','12', '123121', '{', '1', '}']

word_tokenize(punct_tokenize(string=s)) == expected # True
punct_tokenize(string=s).split() == expected  # True
```

Note: If you don't need other fancy regex from the `TreebankWordTokenizer`, the simple `punct_tokenize` should be enough.

**Comment by Alvaro-Kothe** [comment]
From what I've read in this discussion, this bug won't be fixed. I think this issue can be closed.

**Comment by ekaf** [comment]
Thanks @alvations and @Alvaro-Kothe. So there is a [recommended workaround here](https://github.com/nltk/nltk/issues/1963#issuecomment-367185040).
